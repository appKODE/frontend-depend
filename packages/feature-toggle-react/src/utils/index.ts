export { FeatureToggleProvider } from './provider'
export { FeatureToggle } from './feature-toggle-component'
export { useFeatureToggle } from './use-feature-toggle'
