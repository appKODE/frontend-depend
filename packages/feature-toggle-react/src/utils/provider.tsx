import React, { ReactNode, createContext, useEffect, useState } from 'react'

import { FeatureToggleConfig } from '../feature-toggle-config'
import {
  FeatureFlags,
  FeatureToggleContext as FeatureToggleContextType,
} from '../types'

const initialValue = {} as any

export const FeatureToggleContext =
  createContext<FeatureToggleContextType>(initialValue)

type Props = {
  children: ReactNode
  config: FeatureToggleConfig
}

export const FeatureToggleProvider = ({ children, config }: Props) => {
  const {
    defaultFlags,
    fetcher,
    storage: { getFeatureFlags, setFeatureFlags },
  } = config

  const [stateFlags, setStateFlags] = useState(defaultFlags)
  const [isFetching, setIsFetching] = useState(false)

  useEffect(() => {
    const setFlags = async () => {
      const flags = await getFeatureFlags()

      setStateFlags(flags)
    }

    const getRemoteFlags = async () => {
      if (fetcher) {
        try {
          setIsFetching(true)
          const flags = await fetcher()

          if (flags) {
            setStateFlags(prev => {
              const newFlags = {
                ...prev,
                ...flags,
              }

              setFeatureFlags(newFlags)

              return newFlags
            })
          }
        } catch (error) {
          console.error(error)
        } finally {
          setIsFetching(false)
        }
      }
    }

    const init = async () => {
      await setFlags()
      getRemoteFlags()
    }

    init()
  }, [getFeatureFlags])

  const hasFeatureFlag = (flag: keyof FeatureFlags) => {
    return stateFlags[flag]
  }

  const value: FeatureToggleContextType = {
    flags: stateFlags,
    isFetching,
    hasFeatureFlag,
  }

  return (
    <FeatureToggleContext.Provider value={value}>
      {children}
    </FeatureToggleContext.Provider>
  )
}
