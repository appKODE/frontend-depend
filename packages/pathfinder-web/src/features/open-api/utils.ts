export const operationMap = {
  get: 'GET',
  put: 'PUT',
  post: 'POST',
  delete: 'DELETE',
  options: 'OPTIONS',
  head: 'HEAD',
  patch: 'PATCH',
  trace: 'TRACE',
} as const

export const mockSpec = {
  info: {
    version: '1',
    title: 'title',
  },
  servers: [
    {
      url: 'https://stoplight.io/mocks/some/sub-some/123123',
      description: 'cloud mock',
    },
    {
      description: 'local mock',
      url: 'http://185.0.0.1:3100',
    },
    {
      url: 'https://some.someweb.com/api',
      description: 'dev',
    },
  ],
  paths: {
    '/auth/v1/confirm': {
      post: {
        summary: 'verification',
        operationId: 'post-auth-confirm',
        responses: {
          200: {
            description: 'Successful authorization',
            content: {
              'application/json': {
                schema: {
                  title: 'TokensPair',
                  type: 'object',
                  description: 'Tokens Pair',
                  properties: {
                    accessToken: {
                      type: 'string',
                      description: 'Access Token',
                    },
                    refreshToken: {
                      title: 'RefreshToken',
                      type: 'string',
                      description: 'Refresh Token',
                    },
                  },
                  required: ['accessToken', 'refreshToken'],
                },
              },
            },
          },
          403: {
            description: 'attempts amount',
            content: {
              'application/json': {
                schema: {
                  title: 'ConfirmErrorResponse',
                  type: 'object',
                  properties: {
                    attempts: {
                      type: 'integer',
                      format: 'uint64',
                      minimum: 1,
                      exclusiveMinimum: false,
                    },
                  },
                  required: ['attempts'],
                },
              },
            },
            headers: {},
          },
          422: {
            description: 'Example response',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  title: 'ConfirmResponse422',
                  properties: {
                    code: {
                      type: 'string',
                      title: 'ConfirmCode422',
                      enum: [
                        'ErrUserNotFound',
                        'ErrTooManyContacts',
                        'ErrUserHasNotChildren',
                      ],
                    },
                  },
                  required: ['code'],
                },
              },
            },
          },
          500: {
            description: 'Internal Server Error',
          },
        },
        requestBody: {
          content: {
            'application/json': {
              schema: {
                title: 'ConfirmRequest',
                type: 'object',
                properties: {
                  otpID: {
                    type: 'string',
                    title: 'OTPID',
                    description: 'OTP identifier',
                    format: 'uuid',
                  },
                  otpCode: {
                    type: 'string',
                    description: 'OTP code',
                    pattern: '\\d+',
                    minLength: 1,
                  },
                },
                required: ['otpID', 'otpCode'],
              },
            },
          },
        },
        description: '',
        tags: ['auth'],
        'x-internal': false,
        security: [],
      },
      parameters: [],
    },
  },
}
