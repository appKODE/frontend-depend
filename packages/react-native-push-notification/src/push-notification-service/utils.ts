import notifee, {
  AndroidBigPictureStyle,
  AndroidStyle,
  AuthorizationStatus,
  IOSNotificationAttachment,
  InitialNotification,
  TriggerNotification,
} from '@notifee/react-native'
import messaging, {
  FirebaseMessagingTypes,
} from '@react-native-firebase/messaging'
import { Image, Linking, Platform } from 'react-native'

import { TFirebaseRemoteMessage, TPushNotificationBody } from './types'
import { isPermisionsGrantedByDefault } from '../model/constanst'

let initialRemotePushWasGetting = false
let initialLocalPushWasGetting = false

const isPermitted = (authorizationStatus: AuthorizationStatus) => {
  return (
    authorizationStatus === messaging.AuthorizationStatus.AUTHORIZED ||
    authorizationStatus === messaging.AuthorizationStatus.PROVISIONAL
  )
}

/**
 * Checks for permissions to send notifications
 */
export async function hasPermission() {
  const status = await notifee.getNotificationSettings()

  return isPermitted(status.authorizationStatus)
}

/**
 * Request permissions to receive push notifications
 */
export async function requestPermission() {
  const status = await notifee.requestPermission()

  return isPermitted(status.authorizationStatus)
}

/**
 * Checks and request permissions to receive push notifications
 */
export async function checkPermission() {
  const enabledNotifications = await hasPermission()

  if (!enabledNotifications && !isPermisionsGrantedByDefault) {
    return await requestPermission()
  }
  return enabledNotifications
}

/**
 * Generating message id
 */
const createUUIDService = () => {
  let id = 0

  return {
    getUuid: () => {
      return String(++id)
    },
  }
}
const UUIDService = createUUIDService()

/**
 * Receives scheduled local push notifications
 */
export function getScheduledLocalNotifications(): Promise<
  TriggerNotification[]
> {
  return new Promise((resolve, reject) => {
    try {
      const notifications = notifee.getTriggerNotifications()
      resolve(notifications)
    } catch (e) {
      reject(e)
    }
  })
}

/**
 * Cancels all local pushes
 */
export function cancelAllNotifications(): Promise<void> {
  return new Promise(async (resolve, reject) => {
    try {
      const notifications = await getScheduledLocalNotifications()
      const notificationIds = notifications.map(
        item => item.notification.id ?? '',
      )
      resolve(notifee.cancelAllNotifications(notificationIds))
    } catch (e) {
      reject(e)
    }
  })
}

/**
 * Cancels local push by id
 */
export function cancelNotification(notificationId: string): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      resolve(notifee.cancelNotification(notificationId))
    } catch (e) {
      reject(e)
    }
  })
}

/**
 * Gets the initial remote push (from Firebase)
 */
export function getInitialPush(): Promise<FirebaseMessagingTypes.RemoteMessage | null> {
  return new Promise((resolve, reject) => {
    try {
      if (initialRemotePushWasGetting) {
        return resolve(null)
      }
      const initialPush = messaging().getInitialNotification()
      resolve(initialPush)
      initialRemotePushWasGetting = true
    } catch (e) {
      reject(e)
    }
  })
}

/**
 * Gets the initial local push
 */
export function getInitialLocalPush(): Promise<InitialNotification | null> {
  return new Promise((resolve, reject) => {
    try {
      if (initialLocalPushWasGetting) {
        return resolve(null)
      }
      const initialPush = notifee.getInitialNotification()
      resolve(initialPush)
      initialLocalPushWasGetting = true
    } catch (e) {
      reject(e)
    }
  })
}

/**
 * Gets the Firebase FCM-token for push notifications
 */
export const getToken = async () => {
  try {
    const enabledNotifications = await checkPermission()

    if (!enabledNotifications) {
      return null
    }
  } catch (e) {
    return null
  }
  let fcmToken = null
  try {
    fcmToken = await messaging().getToken()
    return fcmToken
  } catch (e) {}
  return fcmToken
}

/**
 * Removes the FCM-token in Firebase. The device will not be able to receive push notifications
 */
export const removeToken = async () => {
  await messaging().deleteToken()
}

/**
 * Go to notification settings
 */
export const openSettings = Platform.select({
  ios: Linking.openSettings,
  android: notifee.openNotificationSettings,
})!

/**
 * Set the number on the badge (iOS)
 */
export const setApplicationIconBadgeNumber = (number: number) => {
  try {
    notifee.setBadgeCount(number)
  } catch (e) {}
}

/**
 * Increases the number on the badge (iOS)
 */
export const incrementApplicationIconBadgeNumber = () => {
  try {
    notifee.incrementBadgeCount()
  } catch (e) {}
}

/**
 * Decreases the number on the badge (iOS)
 */
export const decrementApplicationIconBadgeNumber = () => {
  try {
    notifee.decrementBadgeCount()
  } catch (e) {}
}

/**
 * Retrieves attachment on iOS from remoteMessage
 */
export const getIosAttachment = async (message: TFirebaseRemoteMessage) => {
  if (message.data?.fcm_options?.image) {
    try {
      /**
       * Let's make sure the attachment is available on iOS
       */
      await Image.prefetch(message.data.fcm_options.image)
    } catch (e) {
      delete message.data.fcm_options.image
    }
  }

  const attachment: IOSNotificationAttachment[] = message.data?.fcm_options
    ?.image
    ? [
        {
          url: message.data.fcm_options.image,
        },
      ]
    : []

  return attachment
}

/**
 * Retrieves attachment on Android from remoteMessage
 */
export const getAndroidAttachment = async (
  message: FirebaseMessagingTypes.RemoteMessage,
) => {
  if (message.notification?.android?.imageUrl) {
    try {
      /**
       *  Let's make sure the attachment is available on Android
       */
      await Image.prefetch(message.notification.android.imageUrl)
    } catch (e) {
      delete message.notification.android.imageUrl
    }
  }

  const attachment: AndroidBigPictureStyle | undefined = message.notification
    ?.android?.imageUrl
    ? {
        type: AndroidStyle.BIGPICTURE,
        picture: message.notification.android.imageUrl,
      }
    : undefined

  return attachment
}

export const mapRemotePushToLocalPush = (
  remotePush: FirebaseMessagingTypes.RemoteMessage,
) => {
  const pushNotification: TPushNotificationBody = {
    id: UUIDService.getUuid(),
    title: remotePush.notification?.title ?? '',
    body: remotePush.notification?.body ?? '',
    data: remotePush.data,
  }
  return pushNotification
}
