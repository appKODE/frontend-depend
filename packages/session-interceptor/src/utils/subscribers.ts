import { TStorage, TStorageGetter, TStorageSetter } from './types'

type TSubscriber = () => void

const STORAGE_KEY = 'SESSION_IS_ALLOW_TO_REFETCH'
export class Subscribers {
  constructor({ storageGetter, storageSetter }: TStorage) {
    this.storageGetter = storageGetter
    this.storageSetter = storageSetter

    this.setAllowToRefetch(true)
  }

  storageGetter: TStorageGetter = () => {
    return ''
  }
  storageSetter: TStorageSetter = () => null

  $subscribers: TSubscriber[] = []

  subscribe(callback: () => void) {
    this.$subscribers.push(callback)
  }

  #clear() {
    this.$subscribers = []
  }

  #invokeAll() {
    return this.$subscribers.forEach(cb => cb())
  }

  startInvokes() {
    setInterval(() => {
      if (this.getAllowToRefetch() && this.$subscribers.length > 0) {
        this.#invokeAll()
        this.#clear()
      }
    }, 1000)
  }

  getAllowToRefetch() {
    return this.storageGetter(STORAGE_KEY) === 'true' ? true : false
  }

  setAllowToRefetch(payload: boolean) {
    this.storageSetter(STORAGE_KEY, payload ? 'true' : 'false')
  }
}
