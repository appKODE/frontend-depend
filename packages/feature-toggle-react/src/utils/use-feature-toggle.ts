import { useContext } from 'react'

import { FeatureToggleContext } from './provider'

export const useFeatureToggle = () => useContext(FeatureToggleContext)
