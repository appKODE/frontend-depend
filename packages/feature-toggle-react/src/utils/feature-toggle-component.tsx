import React, { ReactNode } from 'react'

import { FlagsNameProp } from '../types'

import { useFeatureToggle } from './use-feature-toggle'

type Props = {
  active: ReactNode
  inactive: ReactNode
} & FlagsNameProp

export const FeatureToggle = ({ active, inactive, name }: Props) => {
  const { hasFeatureFlag } = useFeatureToggle()

  return <>{hasFeatureFlag(name) ? active : inactive}</>
}
