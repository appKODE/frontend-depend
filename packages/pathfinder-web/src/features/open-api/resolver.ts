import { isOpenApiSpec, isOperationType } from './validators'
import { operationMap } from './utils'
import { ParseResult, Server, UrlSpec } from '../../types'
import { mapResponses } from './mappers'

export function parseJSON(json: unknown): ParseResult {
  if (!isOpenApiSpec(json)) {
    throw new Error('parse error')
  }

  const urls: UrlSpec[] = []

  const urlTemplates = Object.keys(json.paths)
  for (const urlTemplate of urlTemplates) {
    const methods = Object.keys(json.paths[urlTemplate]) as Array<
      keyof typeof operationMap
    >

    for (const method of methods) {
      if (isOperationType(method)) {
        const operation = json.paths[urlTemplate][method]!
        const url: UrlSpec = {
          id: operation.operationId,
          name: operation.summary,
          tags: operation.tags || [],
          template: urlTemplate,
          method: operationMap[method],
          responses: mapResponses(operation.responses || {}),
        }
        urls.push(url)
      }
    }
  }

  return {
    id: json.info.title,
    envs: [
      ...json.servers.map((server: Server, index: number) => ({
        id: index.toString(),
        name: server.description || server.url,
        baseUrl: server.url,
        queryParams: {},
      })),
    ],
    urls,
  }
}
