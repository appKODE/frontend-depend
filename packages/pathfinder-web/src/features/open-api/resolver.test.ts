import { parseJSON } from './resolver'
import { mockSpec } from './utils'

test('snapshots should match', () => {
  expect(parseJSON(mockSpec)).toMatchSnapshot()
})
