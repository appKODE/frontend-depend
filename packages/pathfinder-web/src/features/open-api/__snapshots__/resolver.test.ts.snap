// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match 1`] = `
Object {
  "envs": Array [
    Object {
      "baseUrl": "https://stoplight.io/mocks/some/sub-some/123123",
      "id": "0",
      "name": "cloud mock",
      "queryParams": Object {},
    },
    Object {
      "baseUrl": "http://185.0.0.1:3100",
      "id": "1",
      "name": "local mock",
      "queryParams": Object {},
    },
    Object {
      "baseUrl": "https://some.someweb.com/api",
      "id": "2",
      "name": "dev",
      "queryParams": Object {},
    },
  ],
  "id": "title",
  "urls": Array [
    Object {
      "id": "post-auth-confirm",
      "method": "POST",
      "name": "verification",
      "responses": Array [
        Object {
          "code": "200",
          "examples": Array [],
        },
        Object {
          "code": "403",
          "examples": Array [],
        },
        Object {
          "code": "422",
          "examples": Array [],
        },
        Object {
          "code": "500",
          "examples": Array [],
        },
      ],
      "tags": Array [
        "auth",
      ],
      "template": "/auth/v1/confirm",
    },
  ],
}
`;
